lib/core/test/unit/Cardano/Pool/RankingSpec.hs

```diff
@@ -0,0 +1,254 @@
+                $ property
+                $ allElseEqualProperty @"margin" NegativeEffect
+
+            it "higher cost => lower desirability (or equal)"
+                $ property
+                $ allElseEqualProperty @"cost" NegativeEffect
+
+            it "higher performance => higher desirability (or equal)"
+                $ property
+                $ allElseEqualProperty @"recentAvgPerformance" PositiveEffect
+
+            it "higher oversaturation => desirability NOT affected"
+                $ property prop_oversaturationDoesNotInfluenceDesirability
+
+        describe "desirability" $ do
+            it "desirability >= 0"
+                $ property prop_desirabilityNonNegative
+
+        describe "saturatedPoolRewards" $ do
+            it "a0 == 0 => saturatedPoolRewards == p*R*z0"
+                $ property prop_saturatedPoolRewardsReduces
+
+            it "saturatedPoolRewards >= 0"
+                $ property prop_saturatedPoolRewardsNonNegative
+
+        describe "mkRelativeStake" $ do
+            it "stake > R, R >= 1 => mkRelativeStake ∈ [0,1] (even with conversions)"
+                $ property prop_mkRelativeStakeBounds
+
+prop_mkRelativeStakeBounds :: EpochConstants -> Word64 -> Word64 -> Property
+prop_mkRelativeStakeBounds constants' tot stake =
+    let
+        rel = unRelativeStake
+            $ mkRelativeStake
+                (Lovelace $ fromIntegral stake)
+                constants
+    in
+        counterexample ("relative stake = " ++ show rel) $
+            tot >= 1 ==>
+                tot >= stake ==>
+                    property $ rel >= 0 && rel <= 1
+
+  where
+    constants = constants'
+        { totalRewards = Lovelace . unsafeToNonNegative $ fromIntegral tot }
+
+prop_desirabilityNonNegative :: EpochConstants -> Pool -> Property
+prop_desirabilityNonNegative constants pool = property $
+    desirability constants pool >= 0
+
+prop_saturatedPoolRewardsNonNegative :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsNonNegative constants pool = property $
+    saturatedPoolRewards constants pool >= 0
+
+prop_saturatedPoolRewardsReduces :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsReduces constants' pool =
+    let
+        constants = constants'
+            { leaderStakeInfluence = unsafeToNonNegative 0 }
+        z0 = unRelativeStake  $ saturatedPoolSize constants
+        _R = R.getNonNegative $ getLovelace $ totalRewards constants
+        p  = R.getNonNegative $ recentAvgPerformance pool
+    in
+        saturatedPoolRewards constants pool === p*_R*z0
+
+prop_oversaturationDoesNotInfluenceDesirability
+    :: Pool
+    -> EpochConstants
+    -> Property
+prop_oversaturationDoesNotInfluenceDesirability pool constants =
+    forAll arbitrary $ \v1 v2 -> do
+        let (higherStake, lowerStake) = sortTuple (v1, v2)
+        let z0 = unRelativeStake $ saturatedPoolSize constants
+        higherStake > z0 ==>
```

Comment:
Side note but, this pre-condition should really be `lowerStake > z0`, shouldn't it?

Reply:
If `lowerStake > z0` and `higherStake >= lowerStake`, then both pools would be over-saturated.

It's enough to have the *one* pool be oversaturated, for it to be "more oversaturated" than the other one.

Reply:
But.. the `stake` of the pool doesn't even enter into account in the computation of the desirability. So, the property reduces to: `forall d. d === d ~ True`. This property is a tautology. Let's remove it and feel free to drop a comment in the source module about the effect of over-saturation.

Reply:
> But.. the stake of the pool doesn't even enter into account in the computation of the desirability. So, the property reduces to: forall d. d === d ~ True. This property is a tautology. Let's remove it and feel free to drop a comment in the source module about the effect of over-saturation.

I think my response above applies here. Yes it is a tautology, but the moment we add a `stake` field (as you indirectly suggested in a previous comment, or when we implement non-myopic member rewards), it will not longer be.

Reply:
Although, if you insist, let me know…

There is already a comment on `desirability`.

Reply:
> the moment we add a stake field (as you indirectly suggested in a previous comment, or when we implement non-myopic member rewards), it will not longer be.

Then, we'll add the property at this moment :+1: 
I insist :)